_posts/news/2017-03-21-announcing-scalajs-0.6.15.md

```diff
@@ -0,0 +1,372 @@
+object Foo extends js.Object
+
+@js.native
+@JSGlobal("Foobar")
+class Bar extends js.Object
+{% endhighlight %}
+
+Note that `@JSGlobal` *replaces* `@JSName` if it is present, and is otherwise added.
+
+#### Migration tips
+
+You can temporarily disable the deprecation warnings with the following sbt setting:
+
+{% highlight scala %}
+scalacOptions += "-P:scalajs:suppressMissingJSGlobalDeprecations"
+{% endhighlight %}
+
+This will keep working in the 0.6.x cycle.
+You can use that to ease your migration and delay dealing with these deprecations until you have some time to allocate to that.
+
+You can also use [scalafix](https://scalacenter.github.io/scalafix/) to automatically migrate your codebase!
+TODO Ask @olafurpg about the exact configuration.
+
+### `@JSExport` on objects and classes becomes `@JSExportTopLevel`
+
+To make a top-level class or object accessible to JavaScript, we previously used the `@JSExport` annotation, like this:
+
+{% highlight scala %}
+package foo
+
+@JSExport
+object Foo
+
+@JSExport("Bar")
+class Bar
+{% endhighlight %}
+
+so that they can be accessed from JavaScript as:
+
+{% highlight javascript %}
+var fooObject = foo.Foo(); // note the ()
+var barObject = new Bar();
+{% endhighlight %}
+
+Starting with 0.6.15, `@JSExportTopLevel` should be used instead.
+On *classes*, it has exactly the same behavior as `@JSExport`.
+On *objects* however, it directly exports the object, rather than a 0-arg function.
+Moreover, at least for now, `@JSExportTopLevel` always demands an explicit string argument.
+
+We would therefore update the previous example as follows:
+
+{% highlight scala %}
+package foo
+
+@JSExportTopLevel("foo.Foo")
+object Foo
+
+@JSExportTopLevel("Bar")
+class Bar
+{% endhighlight %}
+
+and use it from JavaScript as:
+
+{% highlight javascript %}
+var fooObject = foo.Foo; // note the absence of ()
+var barObject = new Bar();
+{% endhighlight %}
+
+If retaining the 0-arg function is desirable in your use case (so that you do not need to update the JavaScript code), you can do the following instead:
+
+{% highlight scala %}
+package foo
+
+object Foo {
+  @JSExportTopLevel("foo.Foo")
+  protected def jsAccessor(): this.type = this
+}
+{% endhighlight %}
+
+#### Migration tips
+
+You can temporarily disable the deprecation warnings with the following sbt setting:
+
+{% highlight scala %}
+scalacOptions += "-P:scalajs:suppressExportDeprecations"
+{% endhighlight %}
+
+This will keep working in 0.6.x.
+
+#### About `js.JSApp`
+
+If you use `js.JSApp`, you need not change anything.
+We have seen people do the following (both `extends js.JSApp` *and* `@JSExport`):
+
+{% highlight scala %}
+@JSExport
+object Foo extends js.JSApp {
+  @JSExport
+  def main(): Unit = { ... }
+}
+{% endhighlight %}
+
+This is redundant.
+You can safely remove the two `@JSExport`s while leaving the behavior unchanged:
+
+{% highlight scala %}
+object Foo extends js.JSApp {
+  def main(): Unit = { ... }
+}
+{% endhighlight %}
```

Comment:
`JSApp` still has `@JSExportDescendentObjects`, so the change is safe regardless in 0.6.x. We'll have to decide what to do with `JSApp` with 1.0.0, though, but that's another issue. My current idea is to:
* re-introduce `main(Array[String])` detection in sbt
* use a `ModuleInitializer` that can call those methods
* add `def main(args: Array[String]): Unit = main()` in `js.JSApp` for compatibility
* deprecate `JSApp`